torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -313,7 +313,12 @@ void removeProfileNodesAndSpecializeTypes(Block* b) {
       if (profiled_type == TensorType::get()) {
         continue;
       }
-      it->input()->setType(it->ty(attr::profiled_type));
+      // If we encounter non-identical profiled types for the same value, merge
+      // them.  This situation can happen if, e.g., loop unrolling duplicates
+      // profiled types in a loop body in a manner that isn't logically
+      // consistent (see TestTEFuser.test_unrolled_cat).
+      it->input()->setType(
+          it->input()->type()->expect<TensorType>()->merge(*profiled_type));
```

Comment:
Isn't this always resulting in non-specified type? In what situations `it->input()->type()->expect<TensorType>()` would have a concrete type?

Reply:
I think with this change the fuser might just end up fusing nothing

Reply:
if two types are concrete and match exactly, we will get the complete type.
We would probably end up fusing less for sure. That's pretty much a main reason why this logic was implemented the way it is i.e. we wanted to speculate and pick a concrete type.

Reply:
Oh yeah, this basically disables the fuser.  I mistakenly thought it->input() actually had a type, but it does not, it's always Tensor (except for those rare cases where it is simply wrong).

Reply:
@bertmaher it should probably be : 

```cpp
      if (profiled_type == TensorType::get()) {
        continue;
      }

      if (it->input()->type() == TensorType::get()) {
          it->input()->setType(it->ty(attr::profiled_type));
      }
      else {
          it->input()->setType(it->input()->merge(it->ty(attr::profiled_type)));
      }
```

Reply:
> Oh yeah, this basically disables the fuser. I mistakenly thought it->input() actually had a type, but it does not, it's always Tensor (except for those rare cases where it is simply wrong).

This is exactly what I pointed out in the chat earlier today. Nick's latest fix should handle this case.

Reply:
if you guys really don't trust merge and I can't blame you,
we could try something like this?

```cpp
      if (it->input()->type() == TensorType::get() || *it->ty(attr::profiled_type) == *it->input()->type()) {
          it->input()->setType(it->ty(attr::profiled_type));
      }
```

Note `*` otherwise we would be comparing pointers to types rather then calling the override for ==. 
`it->input()->type() == TensorType::get()` is okay though since there could be only one `TensorType::get()`